### PR TITLE
feat: Added Modular Exponentiation

### PR DESCRIPTION
**Pull request by proRamLOGO**
#### Description of Change
Added Modular Exponentiation iterative and recursive.
Added bitwise operations to speed up the processing of program.

#### Checklist

- [x] Added description of change
- [x] Added file name matches [File name guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#New-File-Name-guidelines)
- [x] Added tests and example, test must pass
- [x] Added documentation so that the program is self-explanatory and educational - [Doxygen guidelines](https://www.doxygen.nl/manual/docblocks.html)
- [x] Relevant documentation/comments is changed or added
- [ ] PR title follows semantic [commit guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#Commit-Guidelines)
- [x] Search previous suggestions before making a new one, as yours may be a duplicate.
- [x] I acknowledge that all my contributions will be made under the project's license.

Notes: Added 2 functions with relevant descriptions and tests.

<a href="https://gitpod.io/#https://github.com/TheAlgorithms/C-Plus-Plus/pull/1411"></a>

### CODE REVIEW

File: math/binary_exponent.cpp [hunk 1]
```diff
@@ -19,53 +19,92 @@
...
+    if (b&1) { // Same as b%2==1. Result to true if b is odd.
         return res * res * a;
     } else {
         return res * res;
     }
 }
 
+/// Recursive function to calculate exponent modular m (a^b %m) in \f$O(\log(n))\f$ using 
+/// binary exponentiation.
+uint64_t binModExpo(uint64_t a, uint64_t b, uint64_t m) {
```
Comment (Panquesito7) [inline math/binary_exponent.cpp:51]: Function documentation is not up to the [Doxygen guidelines](https://www.doxygen.nl/manual/docblocks.html).

File: math/binary_exponent.cpp [hunk 2]
```diff
@@ -19,53 +19,92 @@
...
-        b /= 2;
+        b >>= 1;
+    }
+    return res;
+}
+
+
+/// Iterative function to calculate exponent modular m (a^b %m) in \f$O(\log(n))\f$ using
+/// binary exponentiation.
+uint64_t binModExpo_alt(uint64_t a, uint64_t b, uint64_t m) {
```
Comment (Panquesito7) [inline math/binary_exponent.cpp:80]: Same here as above.

File: math/binary_exponent.cpp [hunk 3]
```diff
@@ -19,53 +19,92 @@
...
+        if (b&1) {
+            res = (res * a) % m ;
+        }
+        a = (a * a) % m;
+        b >>= 1 ;
     }
     return res;
 }
 
 /// Main function
```
Comment (Panquesito7) [inline math/binary_exponent.cpp:93]:
```suggestion
/**
 * @brief Main function
 * @returns 0 on exit
 */
```

File: math/binary_exponent.cpp [hunk 4]
```diff
@@ -19,53 +19,92 @@
...
-        int resRecurse = binExpo(a, b);
-        /// int resIterate = binExpo_alt(a, b);
-
-        /// Result of a^b (where '^' denotes exponentiation)
-        std::cout << resRecurse << std::endl;
-        /// std::cout << resIterate << std::endl;
+        std::cout << "Result of a^b Recursive : "   << binExpo(a,b)          << std::endl;
+        std::cout << "Result of a^b%m Recursive : " << binModExpo(a,b,m)     << std::endl;
+        std::cout << "Result of a^b Iterative : "   << binExpo_alt(a,b)      << std::endl;
+        std::cout << "Result of a^b%m Iterative : " << binModExpo_alt(a,b,m) << std::endl;
```
Comment (Panquesito7) [inline math/binary_exponent.cpp:108]:
```suggestion
        std::cout << "Result of a^b%m Iterative : " << binModExpo_alt(a,b,m) << std::endl;

        return 0;
```

**Review by Panquesito7 — CHANGES_REQUESTED** [review]
Is this PR about adding Modular Exponentiation, or about improving `math/binary_exponent.cpp`?

**Comment by proRamLOGO** [comment]
> Is this PR about adding Modular Exponentiation, or about improving `math/binary_exponent.cpp`?

Adding Modular Exponentiation

**Pull request closed without merging**